File: elixir/lib/2016/03_squares_with_three_sides/aoc201603.ex
defmodule AOC2016.Day03 do
  @moduledoc """
  Advent of Code 2016, day 3: Squares With Three Sides
  """
  require AOC

  @doc """
  Parse input
  """
  def parse(puzzle_input), do: puzzle_input |> String.split("\n") |> Enum.map(&parse_row/1)

  @doc """
  Parse one row of integers

  ## Examples:

      iex> parse_row("1 14 311 ")
      [1, 14, 311]
  """
  def parse_row(row), do: row |> String.split() |> Enum.map(&String.to_integer/1)

  @doc """
  Solve part 1
  """
  def part1(input), do: input |> Enum.count(&is_triangle?/1)

  @doc """
  Solve part 2
  """
  def part2(input), do: input |> transpose() |> Enum.count(&is_triangle?/1)

  @doc """
  Check if the sides can represent a triangle

  Let a, b, c represent the sides of the triangle, such that a <= b <= c. Then
  we want to check that a + b > c, or equivalently, that a + b + c > 2c.

  ## Examples:

      iex> is_triangle?([3, 4, 5])
      true

      iex> is_triangle?([5, 10, 25])
      false

      iex> is_triangle?([25, 10, 5])
      false

      iex> is_triangle?([1, 2, 3])
      false
  """
  def is_triangle?(sides), do: Enum.sum(sides) > 2 * Enum.max(sides)

  @doc """
  Transpose 3x3 blocks in the list of data

  ## Examples:

      iex> transpose([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
      [[1, 4, 7], [2, 5, 8], [3, 6, 9]]
  """
  def transpose(data) do
    data
    |> Enum.chunk_every(3)
    |> Enum.flat_map(fn [[a1, a2, a3], [b1, b2, b3], [c1, c2, c3]] ->
      [[a1, b1, c1], [a2, b2, c2], [a3, b3, c3]]
    end)
  end

  def main(args) do
    Enum.map(args, fn path -> AOC.solve(path, &parse/1, &part1/1, &part2/1) end)
  end
end

File: python/2016/03_squares_with_three_sides/aoc201603.py
"""AoC 3, 2016: Squares With Three Sides"""

# Standard library imports
import itertools
import pathlib
import sys


def parse(puzzle_input):
    """Parse input"""
    return [
        [int(side) for side in triangle.split()]
        for triangle in puzzle_input.split("\n")
    ]


def part1(data):
    """Solve part 1"""
    return sum(is_triangle(side_lengths) for side_lengths in data)


def part2(data):
    """Solve part 2"""
    return sum(is_triangle(side_lengths) for side_lengths in transpose(data))


def is_triangle(side_lengths):
    """Check that the given side lengths represent a valid triangle

    ## Examples:

    >>> is_triangle([3, 4, 5])
    True
    >>> is_triangle([5, 10, 25])
    False
    >>> is_triangle([25, 10, 5])
    False
    >>> is_triangle([1, 2, 3])
    False
    """
    return sum(side_lengths) > 2 * max(side_lengths)


def transpose(triangles):
    """Transpose list of triangles

    ## Example:

    >>> transpose([[1, 2, 3], [4, 5, 6], [7, 8, 9], [0, 1, 2], [3, 4, 5], [6, 7, 8]])
    [[1, 4, 7], [0, 3, 6], [2, 5, 8], [1, 4, 7], [3, 6, 9], [2, 5, 8]]
    """
    side_lengths = itertools.chain(*zip(*triangles))
    return [list(itertools.islice(side_lengths, 3)) for _ in triangles]


def solve(puzzle_input):
    """Solve the puzzle for the given input"""
    data = parse(puzzle_input)
    solution1 = part1(data)
    solution2 = part2(data)

    return solution1, solution2


if __name__ == "__main__":
    for path in sys.argv[1:]:
        print(f"\n{path}:")
        solutions = solve(puzzle_input=pathlib.Path(path).read_text().strip())
        print("\n".join(str(solution) for solution in solutions))
